### ISSUE TITLE
[restrict-plus-operands] False positive when using Extract and generics

### ISSUE THREAD

**Issue by boris-petrov** [typescript-eslint/typescript-eslint#431]
I'm using the latest version of all packages (1.6.0).

```typescript
function x<T>(z: Extract<keyof T, string>): string {
  return z + '__';
}
```

This warns that the addition is unsafe but it shouldn't. It works correctly if it's not `keyof T` but `keyof obj` where `obj` is some object, i.e. generics are not involved.

**Comment by bradzacher** [comment]
Fixed in #440
